src/drivers/mediatek/mt8186/afe-memif.c

```diff
@@ -0,0 +1,701 @@
+}
+#endif
+static int memif_start(struct dma_chan_data *channel)
+{
+	struct afe_memif_dma *memif = dma_chan_get_data(channel);
+
+	tr_info(&memif_tr, "MEMIF:%d start(%d), channel_status:%d", memif->memif_id, channel->index,
+		channel->status);
+
+	if (channel->status != COMP_STATE_PREPARE && channel->status != COMP_STATE_SUSPEND)
+		return -EINVAL;
+
+	channel->status = COMP_STATE_ACTIVE;
+#if TEST_SGEN
+	mt8186_afe_sinegen_enable(TEST_SGEN_ID, 48000, 1);
+#endif
+	/* Do the HW start of the DMA */
+	afe_memif_set_enable(memif->afe, memif->memif_id, 1);
+
+	return 0;
+}
+
+static int memif_release(struct dma_chan_data *channel)
+{
+	struct afe_memif_dma *memif = dma_chan_get_data(channel);
+
+	/* TODO actually handle pause/release properly? */
+	tr_info(&memif_tr, "MEMIF: release(%d)", channel->index);
+
+	if (channel->status != COMP_STATE_PAUSED)
+		return -EINVAL;
+
+	channel->status = COMP_STATE_ACTIVE;
+	afe_memif_set_enable(memif->afe, memif->memif_id, 0);
+#if TEST_SGEN
+	mt8186_afe_sinegen_enable(TEST_SGEN_ID, 48000, 0);
+#endif
+
+	return 0;
+}
+
+static int memif_pause(struct dma_chan_data *channel)
+{
+	struct afe_memif_dma *memif = dma_chan_get_data(channel);
+
+	/* TODO actually handle pause/release properly? */
+	tr_info(&memif_tr, "MEMIF: pause(%d)", channel->index);
+
+	if (channel->status != COMP_STATE_ACTIVE)
+		return -EINVAL;
+
+	channel->status = COMP_STATE_PAUSED;
+
+	/* Disable HW requests */
+	afe_memif_set_enable(memif->afe, memif->memif_id, 0);
+
+	return 0;
+}
+
+static int memif_stop(struct dma_chan_data *channel)
+{
+	struct afe_memif_dma *memif = dma_chan_get_data(channel);
+
+	tr_info(&memif_tr, "MEMIF: stop(%d)", channel->index);
+	/* Validate state */
+	/* TODO: Should we? */
+	switch (channel->status) {
+	case COMP_STATE_READY:
+	case COMP_STATE_PREPARE:
+		return 0; /* do not try to stop multiple times */
+	case COMP_STATE_PAUSED:
+	case COMP_STATE_ACTIVE:
+		break;
+	default:
+		return -EINVAL;
+	}
+	channel->status = COMP_STATE_READY;
+	/* Disable channel */
+	afe_memif_set_enable(memif->afe, memif->memif_id, 0);
+	return 0;
+}
+
+static int memif_copy(struct dma_chan_data *channel, int bytes, uint32_t flags)
+{
+	struct afe_memif_dma *memif = dma_chan_get_data(channel);
+	struct dma_cb_data next = {
+		.channel = channel,
+		.elem.size = bytes,
+	};
+
+	/* TODO XRUN check, update hw ptr */
+	/* TODO TBD Playback first data check */
+
+	/* update user hwptr */
+	if (memif->direction)
+		memif->wptr = (memif->wptr + bytes) % memif->dma_size;
+	else
+		memif->rptr = (memif->rptr + bytes) % memif->dma_size;
+	tr_dbg(&memif_tr, "memif_copy: wptr:%u, rptr:%u", memif->wptr, memif->rptr);
+
+	notifier_event(channel, NOTIFIER_ID_DMA_COPY, NOTIFIER_TARGET_CORE_LOCAL, &next,
+		       sizeof(next));
+
+	return 0;
+}
+
+static int memif_status(struct dma_chan_data *channel, struct dma_chan_status *status,
+			uint8_t direction)
+{
+	struct afe_memif_dma *memif = dma_chan_get_data(channel);
```

Comment:
do we need to do a null ptr check here? or on channel?

Reply:
I'll check `memif`,
about `channel` the caller is "dma_status", the `channel` will not be nullptr
```
static inline int dma_status(struct dma_chan_data *channel, struct dma_chan_status *status, uint8_t direction)
{
    return channel->dma->ops->status(channel, status, direction);
}
```